File: tx_sysvar/block_ingestion.go
package main

import (
	"time"

	pb "github.com/andrew-solarstorm/yellowstone-grpc-client-go/proto"
)

type LocalBlock struct {
	CreatedAt    int64
	Blocktime    int64
	Slot         uint64
	Transactions []*pb.SubscribeUpdateTransactionInfo
}

type BlockBuilder struct {
	ch  chan *LocalBlock
	dec *Decoder
}

func BlockFormer(dec *Decoder) *BlockBuilder {
	svc := &BlockBuilder{
		dec: dec,
		ch:  make(chan *LocalBlock, 100),
	}
	go svc.worker()
	return svc
}

func (b *BlockBuilder) Queue(block *LocalBlock) {
	b.ch <- block
}
func (b *BlockBuilder) worker() {
	for block := range b.ch {
		now := time.Now()
		for _, tx := range block.Transactions {
			if tx == nil {
				continue
			}

			txCtx := TxContext{
				GeyserSentTime:     now,
				ServerReceivedTime: now,
				Slot:               block.Slot,
				BlockTx:            tx,
				BlockTime:          block.Blocktime,
			}
			b.dec.Queue(&txCtx)
		}
	}
}

File: tx_sysvar_js/block_ingestion.js
const { TxContext } = require('./types');

class LocalBlock {
    constructor(slot, blockTime, createdAt, transactions) {
        this.CreatedAt = createdAt;
        this.Blocktime = blockTime;
        this.Slot = slot;
        this.Transactions = transactions;
    }
}

class BlockBuilder {
    constructor(dec) {
        this.dec = dec;
        this.ch = [];
        this.isWorking = false;
    }

    Queue(block) {
        this.ch.push(block);
        this.worker(); 
    }

    async worker() {
        if (this.isWorking) return;
        this.isWorking = true;

        while (this.ch.length > 0) {
            const block = this.ch.shift();
            const now = new Date();

            for (const tx of block.Transactions) {
                if (!tx) continue;

                const txCtx = new TxContext();
                txCtx.GeyserSentTime = now;
                txCtx.ServerReceivedTime = now;
                txCtx.Slot = block.Slot;
                txCtx.BlockTx = tx;
                txCtx.BlockTime = block.Blocktime;

                this.dec.Queue(txCtx);
            }
            // Yield context so javascript runtime resembles go routines
            await new Promise(resolve => setImmediate(resolve));
        }

        this.isWorking = false;
    }
}

module.exports = {
    LocalBlock,
    BlockBuilder
};
